[NNC] Make splitWithTail transform in-place

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #58270 [NNC] Add documentation for splitWith APIs
* #58269 [NNC] Make splitWithMask transform in-place
* **#58268 [NNC] Make splitWithTail transform in-place**

Differential Revision: [D28427228](https://our.internmc.facebook.com/intern/diff/D28427228)

navahgar committed:
[NNC] Make splitWithTail transform in-place

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

navahgar committed:
Update on "[NNC] Make splitWithTail transform in-place"


Differential Revision: [D28427228](https://our.internmc.facebook.com/intern/diff/D28427228)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

ZolotukhinM commented on test/cpp/tensorexpr/test_loopnest.cpp:
```diff
@@ -698,7 +703,7 @@ TEST(LoopNest, SplitWithTailWithLoopOptions) {
   l.setGPUBlockIndex(loops[0], LoopOptions::IDX_Y);
   l.splitWithTail(loops[0], 4, &outer, &inner, &tail);
-  ASSERT_NE(outer, nullptr);
+  ASSERT_EQ(loops[0], outer);
```
Please also add a test checking that loops from the body remain valid after the split as well (see a comment to the second PR in the stack for an example).

navahgar replied:
Will add this test once `IRMutator` and `Substitute` APIs are fixed.

navahgar committed:
Update on "[NNC] Make splitWithTail transform in-place"


Differential Revision: [D28427228](https://our.internmc.facebook.com/intern/diff/D28427228)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

navahgar commented on torch/csrc/jit/tensorexpr/tensorexpr_init.cpp:
```diff
@@ -413,9 +413,9 @@ void initTensorExprBindings(PyObject* module) {
-            return std::make_tuple(outer, inner, tail);
+            For *inner = nullptr, *tail = nullptr;
+            self.splitWithTail(f, factor, &inner, &tail);
+            return std::make_tuple(inner, tail);
```
@ZolotukhinM Does this change require modifications anywhere else?

navahgar replied:
Yeah. I checked those already. They all pass.

ZolotukhinM commented on benchmarks/cpp/tensorexpr/bench_reduce.cpp:
```diff
@@ -266,10 +266,9 @@ BENCHMARK_DEFINE_F(Reduce1D, TeSplitTail)(benchmark::State& state) {
     te::For* mi;
     te::For* tail;
-    loop.splitWithTail(m, kChunkSize, &mo, &mi, &tail);
+    loop.splitWithTail(m, kChunkSize, &mi, &tail);
```
Nit: can remove `mi` and `tail` entirely since they are unused (here and in other places too).

navahgar replied:
Done.

ZolotukhinM reviewed (approved):
👍

navahgar committed:
Update on "[NNC] Make splitWithTail transform in-place"


Differential Revision: [D28427228](https://our.internmc.facebook.com/intern/diff/D28427228)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).

navahgar committed:
Update on "[NNC] Make splitWithTail transform in-place"


Differential Revision: [D28427228](https://our.internmc.facebook.com/intern/diff/D28427228)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28427228).
